hs_core/tasks.py

```diff
@@ -298,24 +302,45 @@ def delete_zip(zip_path):
 
 @shared_task
 def create_temp_zip(resource_id, input_path, output_path, sf_aggregation):
+    """ Create temporary zip file from input_path and store in output_path
+    :param input_path: full irods path of input starting with federation path
+    :param output_path: full irods path of output starting with federation path
+    :param sf_aggregation: if True, include logical metadata files
+    """
     from hs_core.hydroshare.utils import get_resource_by_shortkey
-    if sf_aggregation:
-        pass
     res = get_resource_by_shortkey(resource_id)
-    full_input_path = '{root_path}/{path}'.format(root_path=res.root_path, path=input_path)
-    istorage = IrodsStorage()
+    istorage = res.get_irods_storage()  # invoke federated storage if necessary
+
+    # update metadata files here, in background
+    if res.resource_type == "CompositeResource":
+        if '/data/contents/' in input_path:
+            short_path = input_path.split('/data/contents/')[1]  # strip /data/contents/
+            res.create_aggregation_xml_documents(aggregation_name=short_path)
+        else:  # all metadata included, e.g., /data/*
+            res.create_aggregation_xml_documents()
 
     try:
         if sf_aggregation:
-            name, ext = os.path.splitext(output_path)
-            head, tail = os.path.split(name)
-            out_with_folder = os.path.join(name, tail)
-            istorage.copyFiles(full_input_path, out_with_folder)
-            istorage.copyFiles(full_input_path + '_resmap.xml',  out_with_folder + '_resmap.xml')
-            istorage.copyFiles(full_input_path + '_meta.xml', out_with_folder + '_meta.xml')
-            istorage.zipup(name, output_path)
-        else:
-            istorage.zipup(full_input_path, output_path)
+            # input path points to single file
+            # ensure that foo.zip contains foo, foo/file1, foo/file2, foo/file3
+            # by copying these into a temp subdirectory foo/foo parallel to where foo.zip is stored
+            temp_folder_name, ext = os.path.splitext(output_path)  # strip .zip to get scratch dir
+            head, tail = os.path.split(temp_folder_name)  # tail is unqualified folder name "foo"
+            out_with_folder = os.path.join(temp_folder_name, tail)  # foo/foo is subdir to zip
+            istorage.copyFiles(input_path, out_with_folder)
+            try:
+                istorage.copyFiles(input_path + '_resmap.xml',  out_with_folder + '_resmap.xml')
+            except SessionException:
+                logger.error("cannot copy {}".format(input_path + '_resmap.xml'))

```

Comment:
This change allows an aggregation to be downloaded even if the metadata files are not present... That might be an acceptable behavior, but it's worth pointing out.

Reply:
@sblack-usu I still haven't gotten to the bottom of the issue that certain .xml single-file-aggregaation metadata files don't get generated on demand. See comment above from Pabitra claiming that he fixed this in an orphaned PR. Until I get that fixed, I can't afford to raise a fatal error if the metadata files are missing.

Reply:
@sblack-usu Pabitra's changes don't address the XML generation problem. The problem is that the Django database correctly contains UTF-8 extensions but the metadata conversion seems to cast them to ASCII by mistake.